src/phoenix/experimental/evals/models/openai.py

```diff
@@ -151,11 +217,76 @@ def _build_messages(
     def _verbose_generation_info(self) -> str:
         return f"OpenAI invocation parameters: {self.public_invocation_params}"
 
-    def _generate(self, prompt: str, **kwargs: Dict[str, Any]) -> str:
+    def generate(
+        self, prompts: List[str], instruction: Optional[str] = None, num_consumers: int = 20
+    ) -> List[str]:
+        return asyncio.run(self._generate_async(prompts, instruction, num_consumers=num_consumers))
+
+    async def _generate_async(
+        self,
+        prompts: List[str],
+        instruction: Optional[str] = None,
+        num_consumers: int = 20,
+    ) -> List[str]:
+        printif(self._verbose, f"Generating responses for {len(prompts)} prompts...")
+        if extra_info := self._verbose_generation_info():
+            printif(self._verbose, extra_info)
+        if not is_list_of(prompts, str):
+            raise TypeError(
+                "Invalid type for argument `prompts`. Expected a list of strings "
+                f"but found {type(prompts)}."
+            )
+        outputs = ["unset"] * len(prompts)
+        queue: asyncio.Queue[Tuple[int, str, Optional[str]]] = asyncio.Queue()
+        SENTINEL = object()  # indicates when the queue is done
+        progress_bar = async_tqdm(total=len(prompts), bar_format=TQDM_BAR_FORMAT)
+
+        producer = self._producer(prompts, instruction, queue, num_consumers, SENTINEL)
+        consumers = [
+            asyncio.create_task(self._consumer(queue, outputs, SENTINEL, progress_bar))
+            for _ in range(num_consumers)
+        ]
+
+        await asyncio.gather(producer, *consumers)
+        return outputs
+
+    async def _producer(
+        self,
+        prompts: List[str],
+        instruction: Optional[str],
+        queue: asyncio.Queue[Tuple[int, str, Optional[str]]],
+        num_consumers: int,
+        sentinel: Any,
+    ) -> None:
+        for index, prompt in enumerate(prompts):
+            await queue.put((index, prompt, instruction))
+        # add a sentinel to the queue for each consumer
+        for _ in range(num_consumers):
+            await queue.put(sentinel)
+
+    async def _consumer(
+        self,
+        queue: asyncio.Queue[Tuple[int, str, Optional[str]]],
+        outputs: List[str],
+        sentinel: Any,
+        progress_bar: async_tqdm[Any],
+    ) -> None:
+        while True:
+            if (item := await queue.get()) is sentinel:
+                break
+
+            index, prompt, instruction = item
+            output = await self._generate(prompt=prompt, instruction=instruction)  # type: ignore
+            logger.info(f"Prompt: {prompt}\nInstruction: {instruction}\nOutput: {output}")
+            outputs[index] = output
+            queue.task_done()
```

Comment:
There's no call to `queue.join`. Is this currently doing anything?

Reply:
I think you're right, this is not needed